### PR TITLE
[DNM] rgw: sync plugin to AWS S3

### PR DESCRIPTION
**Pull request by theanalyst**
super wip, just here for some early feedback.
This is just an experimental sync plugin to sync data from RGW to S3, currently it only does a sync of uploads via simple PUT (need to change this to support multipart upload) & deletes to S3.

Requires some more work to make it work properly, to use multipart uploads to S3, preserve object attributes, understand versioning etc.

### CODE REVIEW

File: src/rgw/rgw_sync_module_aws.cc [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#include "rgw_sync_module_aws.h"
+#include "rgw_rest_conn.h"
+#include "rgw_cr_rest.h"
+
+#define dout_subsys ceph_subsys_rgw
+
+static string aws_object_name(const RGWBucketInfo& bucket_info, const rgw_obj_key&key, bool user_buckets=false){
+  string obj_name;
+  if(!user_buckets){
+    obj_name = bucket_info.owner.tenant + bucket_info.owner.id + "/" + bucket_info.bucket.name + "/" + key.name;
```
Comment (yehudasa) [inline src/rgw/rgw_sync_module_aws.cc]: @theanalyst should probably add a separator between owner.tenant to owner.id

**Comment by yehudasa** [comment]
@theanalyst overall it looks pretty good. One major issue is that it doesn't stream objects from source zone upwards, meaning it's not going to handle nicely objects that aren't tiny. At the time I was looking at implementing it, and that was my biggest issue. I think it requires creating a new pipe CR that gets two rest connections, one for input, another for output and splices between them. Can be potentially based on the cr producer-consumer utility, but not necessarily. This will require us to be able to do both streaming reads, and streaming writes. I'm not sure we have CRs for both, will need to create what's missing. As to multipart upload, should probably come after we figured out the stream, then it might be easier to implement.

File: src/rgw/rgw_sync_module_aws.cc (hunk 1 above)
Comment (theanalyst) [inline src/rgw/rgw_sync_module_aws.cc]: ack

File: src/rgw/rgw_cr_rest.h [hunk 2]
```diff
@@ -65,48 +80,74 @@ class RGWReadRESTResourceCR : public RGWSimpleCoroutine {
...
   RGWRESTConn *conn;
   RGWHTTPManager *http_manager;
   string method;
   string path;
   param_vec_t params;
   T *result;
-  S input;
-
+  bufferlist input_bl;
+  bool send_content_length=false;
```
Comment (amitkumar50) [inline src/rgw/rgw_cr_rest.h]: @theanalyst tab left

File: src/rgw/rgw_cr_rest.h [hunk 3]
```diff
@@ -165,8 +238,8 @@ class RGWPutRESTResourceCR : public RGWSendRESTResourceCR<S, T> {
                         const string& _path,
                         rgw_http_param_pair *_params, S& _input, T *_result)
     : RGWSendRESTResourceCR<S, T>(_cct, _conn, _http_manager,
-                            "PUT", _path,
-                            _params, _input, _result) {}
+                                  "PUT", _path,
```
Comment (amitkumar50) [inline src/rgw/rgw_cr_rest.h]: @theanalyst 2 tabs left

**Comment by yehudasa** [comment]
@theanalyst see https://github.com/yehudasa/ceph/commits/wip-rest-write-cr for my branch that builds on top of yours but does streaming uploads.

**Comment by theanalyst** [comment]
superseded by https://github.com/ceph/ceph/pull/18932

**Pull request closed without merging**